test/unit/puppetlabs/services/jruby/jruby_pool_test.clj

```diff
@@ -15,76 +16,132 @@
   (let [pool-size 2
         config        (jruby-testutils/jruby-puppet-config pool-size)
         profiler      jruby-testutils/default-profiler
-        pool-context  (create-pool-context config profiler)
-        pool          (get-pool pool-context)
+        pool-context  (jruby-core/create-pool-context
+                        config profiler jruby-testutils/default-shutdown-fn)
         err-msg       (re-pattern "Unable to borrow JRuby instance from pool")]
-    (with-redefs [core/create-pool-instance! (fn [_] (throw (IllegalStateException. "BORK!")))]
+    (with-redefs [jruby-internal/create-pool-instance! (fn [_] (throw (IllegalStateException. "BORK!")))]
                  (is (thrown? IllegalStateException (jruby-agents/prime-pool! (:pool-state pool-context) config profiler))))
     (testing "borrow and borrow-with-timeout both throw an exception if the pool failed to initialize"
       (is (thrown-with-msg? IllegalStateException
             err-msg
-            (borrow-from-pool pool)))
+            (jruby-core/borrow-from-pool pool-context)))
       (is (thrown-with-msg? IllegalStateException
             err-msg
-            (borrow-from-pool-with-timeout pool 120))))
+            (jruby-core/borrow-from-pool-with-timeout pool-context 120))))
     (testing "borrow and borrow-with-timeout both continue to throw exceptions on subsequent calls"
       (is (thrown-with-msg? IllegalStateException
           err-msg
-          (borrow-from-pool pool)))
+          (jruby-core/borrow-from-pool pool-context)))
       (is (thrown-with-msg? IllegalStateException
           err-msg
-          (borrow-from-pool-with-timeout pool 120))))))
+          (jruby-core/borrow-from-pool-with-timeout pool-context 120))))))
 
-(deftest test-default-pool-size
-  (let [config jruby-testutils/default-config-no-size
-        profiler   jruby-testutils/default-profiler
-        pool       (create-pool-context config profiler)
-        pool-state @(:pool-state pool)]
-    (is (= core/default-pool-size (:size pool-state)))))
+(defn create-pool-context
+  [max-requests]
+  (let [config (-> (jruby-testutils/jruby-puppet-config 1)
+                   (assoc :max-requests-per-instance max-requests))
+        profiler jruby-testutils/default-profiler
+        pool-context (jruby-core/create-pool-context
+                       config profiler jruby-testutils/default-shutdown-fn)]
+    (jruby-agents/prime-pool! (:pool-state pool-context) config profiler)
+    pool-context))
 
+(deftest flush-jruby-after-max-requests
+  (testing "JRuby instance is not flushed if it has not exceeded max requests"
+    (let [pool-context  (create-pool-context 2)
+          instance      (jruby-core/borrow-from-pool pool-context)
+          id            (:id instance)]
+      (jruby-core/return-to-pool instance)
+      (let [instance (jruby-core/borrow-from-pool pool-context)]
+        (is (= id (:id instance))))))
+  (testing "JRuby instance is flushed after exceeding max requests"
```

Comment:
One other variant of this that I think would be good to have is to see that a JRuby instance which is deeper down in the pool can be flushed when it hits its max requests.  For example, do one borrow and, before returning that instance to the pool, do another borrow which is at its max-requests and confirm that the instance from that second borrow is appropriately flushed.